Reimplementing target description concept using DLTI attribute

nhasabni opened the pull request:
and Interfaces. This is a newer implementation of PR https://github.com/llvm/llvm-project/pull/85141 and [RFC](https://discourse.llvm.org/t/rfc-target-description-and-cost-model-in-mlir/76990) by considering reviews and comments on the original PR.

As an example of attributes supported by this commit:

```
module attributes {
   dlti.target_system_spec =
     #dlti.target_device_spec<
       #dlti.dl_entry<"dlti.device_id", 0: ui32>,
       #dlti.dl_entry<"dlti.device_type", "CPU">,
       #dlti.dl_entry<"dlti.L1_cache_size_in_bytes", 8192 : ui32>>,
     #dlti.target_device_spec <
       #dlti.dl_entry<"dlti.device_id", 1: ui32>,
       #dlti.dl_entry<"dlti.device_type", "GPU">,
       #dlti.dl_entry<"dlti.max_vector_op_width", 64 : ui32>>,
    #dlti.target_device_spec <
       #dlti.dl_entry<"dlti.device_id", 2: ui32>,
       #dlti.dl_entry<"dlti.device_type", "XPU">>>
}
```

nhasabni committed:
Reimplementing target description concept using DLTI attribute

nhasabni commented:
@rengolin @joker-eph Please take a look. This is a DLTI based implementation for representing a target description as attributes. I've also added L1 cache size property and have used this property to calculate block size info in BlockMatMul pass.

rengolin commented:
This is a continuation of #91670 which was unintentionally closed. It addresses @joker-eph's point by adding target properties, not compiler pass options. It also has some usage on different places. 

@adam-smnk

rengolin commented on mlir/lib/Transforms/Canonicalizer.cpp:
```diff
@@ -59,6 +75,41 @@ struct Canonicalizer : public impl::CanonicalizerBase<Canonicalizer> {
+        config.maxNumRewrites = *v;
+      }
+    }
+    LLVM_DEBUG(llvm::dbgs() << "[CostModel] Canonicalizer MaxIterations (new):"
```
This is still here. We don't want any pass information here, just target info.

rengolin commented on mlir/lib/Dialect/Linalg/Transforms/BlockPackMatmul.cpp:
```diff
@@ -134,6 +136,24 @@ transposePackedMatmul(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+    L1_cache_size = *v;
+  }
+
+  // block_size = sqrt(L1_cache_size) rounded down to nearest power of 2.
```
Naively, we need to keep An, Bn and Cn in cache, so probably this should look into 1/3 of that. Better implementations can stream An/Bn, so can have larger tile sizes, but this is just a default anyway.

ftynse commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.h:
```diff
@@ -23,11 +23,17 @@
 // forward declaration, and we need the typedef in the actual declaration.
 using DataLayoutEntryList = llvm::SmallVector<DataLayoutEntryInterface, 4>;
 using DataLayoutEntryListRef = llvm::ArrayRef<DataLayoutEntryInterface>;
+// using TargetDeviceDescSpecList =
+// llvm::SmallVector<TargetDeviceDescSpecInterface, 4>;
```
Let's not commit commented-out code.

nhasabni replied:
Fixed it.

ftynse commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.h:
```diff
@@ -84,6 +90,24 @@ Attribute getDefaultGlobalMemorySpace(DataLayoutEntryInterface entry);
 /// DataLayoutInterface if specified, otherwise returns the default.
 uint64_t getDefaultStackAlignment(DataLayoutEntryInterface entry);
 
+/// return max vector op width from the specified DataLayoutEntry. If the
```
Nit: please capitalize the first word in the sentence.

nhasabni replied:
Fixed it.

ftynse commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.h:
```diff
@@ -84,6 +90,24 @@ Attribute getDefaultGlobalMemorySpace(DataLayoutEntryInterface entry);
+std::optional<int64_t>
+getCanonicalizerMaxIterations(DataLayoutEntryInterface entry);
+
+/// returncanonicalizer max num rewrites from the specified DataLayoutEntry.
```
Nit: something went wrong with the comment here.

nhasabni replied:
Fixed it.

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/BlockPackMatmul.cpp:
```diff
@@ -134,6 +136,24 @@ transposePackedMatmul(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
   return packTransposedMatmul;
 }
 
+static SmallVector<int64_t> getDefaultBlockFactors(linalg::LinalgOp linalgOp) {
```
Could we slice this into a different PR to simplify reviewing?

adam-smnk replied:
+1

ftynse reviewed (changes requested):
Thanks! This generally goes in the right direction, let's iterate on this.

One large comment: this needs standalone tests for the target description part. For the existing DLTI part, we have a test pass that looks at certain kinds of ops and/or attributes and "interprets" them as requests to the data layout object, the results of these requests are appended back to ops as more attributes that we can FileCheck on. Something similar should be possible here.

joker-eph commented on mlir/include/mlir/Dialect/DLTI/DLTI.h:
```diff
@@ -124,6 +126,150 @@ class DataLayoutSpecAttr
+  StringAttr getCanonicalizerMaxIterationsIdentifier(MLIRContext *context);
+
+  /// Returns canonicalizer max num rewrites identifier.
+  StringAttr getCanonicalizerMaxNumRewritesIdentifier(MLIRContext *context);
```
As mentioned in the other PR, please remove the things that do not belong to a "target description".

nhasabni replied:
Fixed it.

joker-eph commented on mlir/include/mlir/Dialect/DLTI/DLTI.h:
```diff
@@ -124,6 +126,150 @@ class DataLayoutSpecAttr
+  getSpecForCanonicalizerMaxNumRewrites(MLIRContext *context);
+
+  /// Return the value of device ID
+  uint32_t getDeviceID(MLIRContext *context);
```
I don't quite get why all these APIs takes a MLIRContext pointer: the Attribute should have already a `getContext()` method.

nhasabni replied:
Fixed it.

joker-eph commented on mlir/include/mlir/Dialect/DLTI/DLTI.h:
```diff
@@ -124,6 +126,150 @@ class DataLayoutSpecAttr
+
+/// A system description attribute is a list of device descriptors, each
+/// having a unique device ID
+class TargetSystemDescSpecAttr
```
As mentioned in the other PR: I don't quite get why these can't be defined in ODS, so please define them in ODS or document why they cannot be.

nhasabni replied:
Fixed it.

joker-eph commented on mlir/include/mlir/Dialect/DLTI/DLTIBase.td:
```diff
@@ -27,6 +27,13 @@ def DLTI_Dialect : Dialect {
+    kTargetSystemDescAttrName = "dlti.tsd_spec";
+
+    constexpr const static ::llvm::StringLiteral
+    kTargetDeviceDescAttrName = "dlti.tdd_spec";
```
The name seems a bit obscure, can we expand the acronym a bit?

nhasabni replied:
Fixed it.

joker-eph commented on mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp:
```diff
@@ -111,6 +113,15 @@ struct RawBufferOpLowering : public ConvertOpToLLVMPattern<GpuOp> {
       uint32_t totalBits = elemBits * dataVector.getNumElements();
+      uint32_t maxVectorOpWidth = 128; // default value
+      if (std::optional<uint32_t> v =
+              DataLayout(gpuOp->template getParentOfType<mlir::ModuleOp>())
```
Is this safe to always assume there is a containing ModuleOp?  Seems to me that this could be a GPUModuleOp for example.

nhasabni replied:
Fixed it.

joker-eph commented on mlir/lib/Dialect/DLTI/DLTI.cpp:
```diff
@@ -337,6 +344,320 @@ void DataLayoutSpecAttr::print(AsmPrinter &os) const {
+constexpr const StringLiteral
+    mlir::DLTIDialect::kTargetDeviceCanonicalizerMaxIterationsKey;
+constexpr const StringLiteral
+    mlir::DLTIDialect::kTargetDeviceCanonicalizerMaxNumRewritesKey;
```
Nit: these aren't needed since C++17 I believe.

nhasabni replied:
Fixed it.

joker-eph commented:
> This is a continuation of https://github.com/llvm/llvm-project/pull/91670 which was unintentionally closed.

FYI: a PR can be reopened after being closed (not merged) instead of opening a new one.

rengolin commented:
> > This is a continuation of #91670 which was unintentionally closed.
> 
> FYI: a PR can be reopened after being closed (not merged) instead of opening a new one.

I tried this but the PR told me the branch was clean and couldn't reopen, even thought it wasn't. I didn't want to fight Github on this. :(

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/BlockPackMatmul.cpp:
```diff
@@ -134,6 +136,24 @@ transposePackedMatmul(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
+static SmallVector<int64_t> getDefaultBlockFactors(linalg::LinalgOp linalgOp) {
+  // get L1 cache size first.
+  uint32_t L1_cache_size = 4096; // default value
+  uint32_t cpuID = 0;
```
nit: defaulting to device 0 is fine but I wouldn't assume it's a CPU

rengolin replied:
I think it's just a naming issue, this should be `deviceID` or something. But also, I wouldn't assume the first one has it either. Perhaps a helper that, when not provided a device ID, returns a `pair<value, id>` if any device has it.

nhasabni replied:
Yes, will fix the naming issue.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/BlockPackMatmul.cpp:
Could you also add a corresponding `.mlir` test that covers the new defaults?

nhasabni committed:
Addressing review comments

1. Use ODS framework for all of DLTI attrs
2. Removing need of MLIRContext in APIs
3. Removing canonicalizer heuristics from this PR

nhasabni commented on mlir/include/mlir/Dialect/DLTI/DLTI.h:
```diff
@@ -124,6 +126,150 @@ class DataLayoutSpecAttr
+// TargetDeviceDescSpecAttr
+//===----------------------------------------------------------------------===//
+
+class TargetDeviceDescSpecAttr
```
Fixed it.

nhasabni commented:
@joker-eph @ftynse Thanks for your comments. I've fixed most of them (incl ODS based implementation). Test cases is WIP. I will also separate out the integration of target properties into the passes into a separate PR.

joker-eph commented on mlir/lib/Transforms/Canonicalizer.cpp:
```diff
@@ -23,6 +23,8 @@ namespace mlir {
 using namespace mlir;
 
+#define DEBUG_TYPE "canonicalizer"
+
```
This can go away now

nhasabni committed:
Addressing review comments; adding unit tests

nhasabni commented:
@joker-eph @ftynse I've addressed all of your comments. Pls take a look. Thanks.

rengolin commented:
Windows bot fails with system error, ignore:
```
fatal error C1060: compiler is out of heap space
```

nhasabni commented:
> Windows bot fails with system error, ignore:
> 
> ```
> fatal error C1060: compiler is out of heap space
> ```

@rengolin Not sure if this is related to the PR. All the unit tests seem to pass.

rengolin commented:
> > Windows bot fails with system error, ignore:
> > ```
> > fatal error C1060: compiler is out of heap space
> > ```
> 
> @rengolin Not sure if this is related to the PR. All the unit tests seem to pass.

It is not, just ignore. The Windows builder is super slow and unreliable.

nhasabni commented:
@ftynse @joker-eph Just a gentle ping for the review.

banach-space commented on mlir/include/mlir/Dialect/DLTI/DLTIAttrs.td:
```diff
@@ -0,0 +1,303 @@
+        Each device description specification describes a single device and
+        its hardware properties. Each device description must have a device_id
+        and a device_type. In addition, the description can contain any number
+        of optional hardware properties (e.g., max_vector_op_width below).
```
Shouldn't optional properties be marked as e.g. `#dlti.dl_optional_entry` rather than `#dlti.dl_entry`? That would make things easier for humans.

nhasabni replied:
That's an interesting thought. So far, the PR only mandates `device_id` and `device_type` properties. Rest all are optional. As the distinction is clear so far, I did not feel the need to create a separate data layout entry type. But if we all feel the need (or as the number of supported mandatory properties grows), we can create an optional property.

banach-space commented on mlir/test/Dialect/DLTI/invalid.mlir:
```diff
@@ -90,3 +90,132 @@ module attributes { dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<"unknown.unknown
+
+// -----
+
+// expected-error@below {{'dlti.target_system_desc_spec' is expected to be a #dlti.target_system_desc_spec attribute}}
```
```suggestion
// expected-error@below {{'dlti.target_system_desc_spec' is expected to be a #dlti.target_device_desc_spec attribute}}
```

Or did I misunderstand this? Though the next test suggests that I got it right :)

rengolin replied:
No, this is just testing the error message for the `42` not being an actual attribute.

banach-space commented on mlir/test/Dialect/DLTI/invalid.mlir:
```diff
@@ -90,3 +90,132 @@ module attributes { dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<"unknown.unknown
+// -----
+
+module attributes {
+  // expected-error@+2 {{target_device_desc_spec requires value of key: dlti.max_vector_op_width to be of ui32 type}}
```
Where is it specified that `dlti.max_vector_op_width ` has to be `ui32`?

nhasabni replied:
Currently, `TargetDeviceDescSpecAttr::verify` routine for `target_device_desc_spec` will enforce the constraint. But I assume this will be documented in a user-facing documentation.

rengolin replied:
I don't see how we can make this a generic documentation when adding new properties without having to change the table-gen description also. For now I'd say let's keep it simple and not worry about it (ie. let compiler developers see the error and fix it). The long term plan is to make this much simpler by having more robust target information structures that don't need to expand all of their properties in a map, so this issue will become irrelevant when actual users start to rely on this.

banach-space reviewed (commented):
Sorry, I've been OOO and just didn't have the cycles to go over this. I've left a few fly-by comments. Should be able to scan the whole thing in the next couple of days.

> We need this implementation agreed and merged so we can start adding actual properties and change the upstream passes to use them, and show the real value from this work.

Do you have a follow-up prototype working that would demonstrate how to use this logic in practice? And how it integrates with Linalg's logic? Why not share it as a draft?

rengolin commented:
> Do you have a follow-up prototype working that would demonstrate how to use this logic in practice?

We had a commit that was using it in the packing pass but removed upon request to keep this PR short.

We have the original [RFC](https://discourse.llvm.org/t/rfc-target-description-and-cost-model-in-mlir/76990), where we discuss the prototypes and ideas. However, after numerous discussions at EuroLLVM and later, we quickly realised that the "cost model" part is really diverse and pre-existing.

Therefore, the focus here is just to describe targets in a reasonable way that any pass can just grab the data and use it internally (as exposed by examples here and the commit that we moved into the next stage). By getting the infrastructure here simple enough, we can start using elsewhere into existing cost models.

So for now this should be used directly, as is, by passes and transforms and third-party cost models. We'll likely change some of the design, this is why we don't have a full pipeline of PRs in the works. It would be a waste of time for all of us. Once we consolidate this implementation, we'll reopen the _upstream_ cost model discussion once again.

> And how it integrates with Linalg's logic?

This has **nothing** to do with Linalg directly. 

Linalg is a dialect that implements linear algebra interfaces, which expose tensor logic to the compiler. Target descriptions just carry information about the hardware and the system. Cost models use that information (plus the state of the IR, and with it, Linalg) to derive answers to common questions. Transforms, rewrites, passes are the ones asking questions, and they are the ones doing anything to Linalg.

ftynse commented on mlir/lib/Dialect/DLTI/DLTI.cpp:
```diff
@@ -331,12 +418,15 @@ DataLayoutSpecAttr DataLayoutSpecAttr::parse(AsmParser &parser) {
 }
 
+#define GET_ATTRDEF_CLASSES
+#include <mlir/Dialect/DLTI/DLTIAttrs.cpp.inc>
```
```suggestion
#include "mlir/Dialect/DLTI/DLTIAttrs.cpp.inc"
```

ftynse commented on mlir/lib/Dialect/DLTI/DLTI.cpp:
```diff
@@ -375,32 +465,35 @@ class TargetDataLayoutInterface : public DataLayoutDialectInterface {
-  if (parser.parseKeyword(&attrKind))
-    return {};
+namespace {
+class SystemDescSpecInterface : public DataLayoutDialectInterface {
```
Document this please.

IIUC, this interface is only meant to run the verifier. This is rather unconventional as one typically wants to have access to the interface. Have you considered other approaches to trigger verification?

ftynse commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -744,6 +820,55 @@ LogicalResult mlir::detail::verifyDataLayoutSpec(DataLayoutSpecInterface spec,
+      if (auto type = llvm::dyn_cast_if_present<Type>(entry.getKey())) {
+        // tdd_spec does not support Type as a key.
+        return failure();
+      } else
+        device_desc_keys[entry.getKey().get<StringAttr>()] = entry;
```
```suggestion
      } else {
        device_desc_keys[entry.getKey().get<StringAttr>()] = entry;
      }
```

ftynse commented on mlir/test/Dialect/DLTI/invalid.mlir:
```diff
@@ -90,3 +90,132 @@ module attributes { dlti.dl_spec = #dlti.dl_spec<#dlti.dl_entry<"unknown.unknown
+      #dlti.dl_entry<"dlti.device_id", 0: ui32>,
+      #dlti.dl_entry<"dlti.device_type", "CPU">>
+  >} {}
+  
```
Please add a trailing newline (make sure your editor doesn't drop it).

ftynse commented on mlir/test/Dialect/DLTI/roundtrip.mlir:
```diff
@@ -53,3 +53,29 @@
+
+// A valid target system description
+// CHECK: module attributes {
+// CHECK-SAME: dlti.target_system_desc_spec = #dlti.target_system_desc_spec<
+// CHECK-SAME:  #dlti.target_device_desc_spec<
+// CHECK-SAME:   #dlti.dl_entry<"dlti.device_id", 0 : ui32>,
+// CHECK-SAME:   #dlti.dl_entry<"dlti.device_type", "CPU">,
+// CHECK-SAME:   #dlti.dl_entry<"dlti.L1_cache_size_in_bytes", 4096 : ui32>>,
+// CHECK-SAME: #dlti.target_device_desc_spec<
+// CHECK-SAME:  #dlti.dl_entry<"dlti.device_id", 1 : ui32>,
+// CHECK-SAME:  #dlti.dl_entry<"dlti.device_type", "GPU">,
+// CHECK-SAME:  #dlti.dl_entry<"dlti.max_vector_op_width", 128 : ui32>>
+// CHECK-SAME: >} {
```
Let's not overconstrain the tests, it's okay for these to be just `CHECK`, there is no fundamental requirement to not have a line break in the syntax.

ftynse commented on mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp:
```diff
@@ -91,6 +94,65 @@ struct CustomDataLayoutSpec
   }
 };
 
+class TargetSystemDescSpecStorage : public AttributeStorage {
```
Please document all top-level objects.

ftynse commented on mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp:
```diff
@@ -308,6 +380,47 @@ struct DLTestDialect : Dialect {
   }
 };
 
+struct DLTargetSystemDescTestDialect : Dialect {
```
```suggestion
struct DLTargetSystemDescTestDialect : public Dialect {
```

ftynse reviewed (changes requested):
LGTM in principle, please address the review comments unless marked non-blocking / for future PR.

ftynse commented:
The original implementation may have gone a bit overboard with two things:
1. Having named methods on the `DataLayout` object doesn't scale very well to adding new kinds of queries.
2. Specialization to Module / builtin types is intrusive.

I don't have a good solution to 1., maybe experiment with more generic methods like `TargetInfo::getIntegerTargetInfoProperty(StringAttr/TypeID/int64_t identifier)` that are parameterized by, e.g., the name of the DLEntry being accessed. This won't necessarily work for queries that require some interpolation logic like integer bitwidth, but will work for discrete cases, which may end being more numerous.

The solution to 2. is to use delayed interface registration + promised interfaces, then drop the special-casing and always go through interfaces. I don't remember if we had that mechanism back when DLTI was introduced, but there was a concern about inefficiency of going through interfaces. The overhead is minor, and we shouldn't optimize prematurely.

nhasabni commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -194,6 +194,162 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
+    size could be a device property, and its value would be a device specific size.
+
+    A target device description specification is attached to a module as a module level
+    attribute.
```
Can you elaborate on this scenario a bit more? Our thought process currently is that the system desc/device desc would contain properties of the underlying hardware (and hence would suffice to be at the module level). Current implementation is not supporting the idea of cost model (or in other words, the heuristics used by compiler passes). Given that which types of properties would be specialized at a function level?

joker-eph replied:
A function could be specialized to use AVX and another for SSE for example.

Look also at the Datalayout analysis in-tree, it kind of achieve this.

rengolin replied:
The way I was thinking this is not to replicate the system descriptor per function, but to have a single (whole) descriptor per module and then an annotation on the op (incl. function) with the identifier (ID or search query) into the module's descriptor.

```mlir
module {
  target = #dlti.system_desc { target[0] = { ... }, target[1] { ... } ... }

  ...
  func.func myFunction(...) #target = 0 {
    ...
  }
  ...
  func.func otherFunc (...) {
    linalg.generic ... #target = 1 {
      ...
    }
  }
  ...
}
```

rengolin replied:
@joker-eph This is the last unaddressed comment. Does my answer help with your question? I think we can solve this later, either relaxing the constraint (as you propose), or creating a hook back to the module's attribute (as I propose). I'd like to get this merged, so we can multi-task the following changes that all depend on this one.

ftynse replied:
Since attributes are owned by context anyway, it doesn't cost us much to allow them at all levels. It would also avoid unchecked indirection (what if someone decides to reorder the list in system description, for example?)

rengolin replied:
Ok, I think we need to have a better API for this anyway, but this PR is already too long to get there. We can have as fewer restrictions as we can right now to not make that harder in the near future.

nhasabni commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -194,6 +194,162 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
+      /*retTy=*/"::mlir::DataLayoutEntryListRef",
+      /*methodName=*/"getEntries",
+      /*args=*/(ins)
+    >,
```
As DeviceDescSpec is conceptually similar to DLTISpec (both have a list of DLTI entries), I was reusing the definitions from DLTISpec.

joker-eph replied:
Basically, a simpler model would look like:

```
#gpu = #my.gpu_desc<{ some_key = some_value, some_other_key = some_value }>
#system = #dlti.target_system_spec<
  devices = {
    "gpu0" = #gpu,
    "gpu1" = #gpu
  },
  system_key = some_value
>
```

nhasabni replied:
> What is a DLTISpec?
> 
> I'm also not really understanding the answer, so I'll ask again: if everything is conceptually a key-value pair, why aren't we using a map-like API?

By DLTISpec I mean [DataLayoutSpecAttr](https://github.com/llvm/llvm-project/blob/ae8627809076390dbab04e01f3bf9d384c9e124e/mlir/include/mlir/Dialect/DLTI/DLTI.h#L69), which is basically a list of DataLayoutEntryAttr.

nhasabni replied:
> Basically, a simpler model would look like:
> 
> ```
> #gpu = #my.gpu_desc<{ some_key = some_value, some_other_key = some_value }>
> #system = #dlti.target_system_spec<
>   devices = {
>     "gpu0" = #gpu,
>     "gpu1" = #gpu
>   },
>   system_key = some_value
> >
> ```
> 
> You can iterate the devices (and they each have a unique id in the system, guaranteed by the key-value pair dictionary), but the device description is orthogonal to the ID which is a property of the system (it just associate a device to a place in the system)

Got it. Would the API to query the value of a property then accept string for the device ID (instead of integer as it is currently)?

joker-eph replied:
StringAttr seems fine?

nhasabni commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -194,6 +194,162 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
+      /*retTy=*/"::mlir::StringAttr",
+      /*methodName=*/"getMaxVectorOpWidthIdentifier",
+      /*args=*/(ins)
+    >,
```
Removed these models from Interface. Yes, these methods are implementation detail that an interface should not contain.

nhasabni commented on mlir/include/mlir/Dialect/DLTI/DLTIAttrs.td:
```diff
@@ -0,0 +1,303 @@
+    /// Returns the stack alignment identifier.
+    StringAttr getStackAlignmentIdentifier(MLIRContext *context) const;
+  }];
+  let extraClassDefinition = [{
+    StringAttr
+    $cppClass::getEndiannessIdentifier(MLIRContext *context) const {
+      return Builder(context).getStringAttr(DLTIDialect::kDataLayoutEndiannessKey);
+    }
+
+    StringAttr
+    $cppClass::getAllocaMemorySpaceIdentifier(MLIRContext *context) const {
+      return Builder(context).getStringAttr(
+          DLTIDialect::kDataLayoutAllocaMemorySpaceKey);
+    }
+
+    StringAttr $cppClass::getProgramMemorySpaceIdentifier(
+        MLIRContext *context) const {
+      return Builder(context).getStringAttr(
+          DLTIDialect::kDataLayoutProgramMemorySpaceKey);
+    }
+
+    StringAttr
+    $cppClass::getGlobalMemorySpaceIdentifier(MLIRContext *context) const {
+      return Builder(context).getStringAttr(
+          DLTIDialect::kDataLayoutGlobalMemorySpaceKey);
+    }
+
+    StringAttr
+    $cppClass::getStackAlignmentIdentifier(MLIRContext *context) const {
+      return Builder(context).getStringAttr(
+          DLTIDialect::kDataLayoutStackAlignmentKey);
+    }
+  }];
```
We wanted to keep decls as wells as defs together rather than spreading over multiple files. But will move those to C++ file.

nhasabni committed:
Addressing review comments

    Mainly as follows:
    - removing custom attribute parser/printer
    - renaming SystemDescSpec to SystemSpec and TargetDeviceDescSpec to
      TargetDeviceSpec
    - using DeviceID as a type insted of uint32_t
    - grammatical errors

joker-eph commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -194,6 +194,162 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
+    Attribute interface describing a target device description specification.
+
+    A target device description specification is a list of device properties (key)
+    and their values for a specific device. The device is identified using "device_id"
```
I don't understand why this is needed though, and why is it exposed here.
If I have a system with 8 identical GPUs, then this interface could be shared between all the GPUs.

fabianmcg commented:
@nhasabni great work!
Quick couple of comments/questions:

1. How do you suggest we model devices of devices (clusters) and so on? 
With the current setup its not entirely evident to me how to model them (maybe I got something wrong in the implementation).

2. Why is the device ID mandatory?
I get its usefulness, for example, describing a system down to its NUMA topology. However, not all system descriptions will care about that, and in those instances it might be highly inconvenient having it mandatory, as systems with the same devices but with different device ids are not equal. I would consider making it optional.

3. Do we want to have methods like `getSpecForL1CacheSizeInBytes` hardcoded in the attribute/interface?
If we have L1, why not L2, l3? I would argue those queries belong to the entries of the spec and not in the attribute.

4. Shouldn't systems have entries as well? Or let me rephrase it, how would you model the amount of shared HBM in a system, is it a device?

nhasabni commented:
Thanks @fabianmcg for your thoughts. See my answers below.
 
> 1. How do you suggest we model devices of devices (clusters) and so on?
>    With the current setup its not entirely evident to me how to model them (maybe I got something wrong in the implementation).

This PR does not discuss the representation of devices. It is meant to get the boiler plate infra support that we envision will be enhanced further. That being said, we did briefly talk about the hierarchical representation of devices (similar to https://discourse.llvm.org/t/representing-device-data/2078/4). Also, [our RFC](https://discourse.llvm.org/t/rfc-target-description-and-cost-model-in-mlir/76990) shades some light on this point.

> 2. Why is the device ID mandatory?
>    I get its usefulness, for example, describing a system down to its NUMA topology. However, not all system descriptions will care about that, and in those instances it might be highly inconvenient having it mandatory, as systems with the same devices but with different device ids are not equal. I would consider making it optional.

It is mandatory currently because the passes (or other users) can identify a particular device using a device ID in order to query its properties. This is because we envision a single instance of system description (as against a device specific description) for the duration of execution.

> 3. Do we want to have methods like `getSpecForL1CacheSizeInBytes` hardcoded in the attribute/interface?
>    If we have L1, why not L2, l3? I would argue those queries belong to the entries of the spec and not in the attribute.

Yes, we would certainly need similar APIs for L2, L3, etc. The point of this PR is not to list all of the possible APIs, but to agree upon the infra to extend it further as needed. Not sure I understand fully what you mean by "queries belong to entries of spec". How would the property query API look like? We were modeling the current API similar to existing API in DLTI dialect.

> 4. Shouldn't systems have entries as well? Or let me rephrase it, how would you model the amount of shared HBM in a system, is it a device?

Yes, this is currently missing. I am working on updating the system description to contain entries as well.

fabianmcg commented:
Thanks for your responses @nhasabni !

> This PR does not discuss the representation of devices. It is meant to get the boiler plate infra support that we envision will be enhanced further.

My question is not about the current PR, but instead about how would one represent those systems with the boiler plate infra. Because from what I understood, right now it's a flat representation, so it works well for a single heterogeneous node, but more complex systems are not straightforward, and we should get that right from the start.


>  It is mandatory currently because the passes (or other users) can identify a particular device using a device ID in order to query its properties.

My comment is about the case: `system1 = [deviceX[id = 0], deviceX[id = 1]]`,  `system2 = [deviceX[id = 1], deviceX[id = 0]]`. In many cases `system1 == system2`, however, they are not the same by forcing ids. Moreover, it's also fragile, because a pass that works on `system1` might not work on `system2` because of the ordering.

Also, why do you need the ID to be stored in the device to query the properties? As far as I could tell, devices are placed in an array, so `attr.getEntries()[devId]` should work.


> Yes, we would certainly need similar APIs for L2, L3, etc.
My issue here is that this doesn't scale. Instead, the generic device attribute should only use a key lookup to find the value. If someone wants to create a more specialized attribute that contains L1, L2... they should be the ones adding those methods.


> Yes, this is currently missing. I am working on updating the system description to contain entries as well.

Cool!

nhasabni committed:
Addressing review comments

Representing TargetSystemSpec as a set of key-value pairs where key is
the DeviceID (string) and the value is TargetDeviceSpec.

rengolin commented:
> My question is not about the current PR, but instead about how would one represent those systems with the boiler plate infra. Because from what I understood, right now it's a flat representation, so it works well for a single heterogeneous node, but more complex systems are not straightforward, and we should get that right from the start.

This is the next step. We have two proposals:
1. **Expand the dictionary representation**. We can make `#dlti.dl_entry` to contain another and cascade sub-ti, we can add arbitrary search inside the tree and have targets of targets represented in a single `#dlti.dl_entry` (for example: getTarget("GPU", "A100", "3") would get the third node in the "A100" sub-entry below the "GPU" top-entry).
2. **Add a generic entry type**. This means `#dlti.dl_entry` accepts a _"descriptor"_, which is an element in a compiler map that can access target information. This allows us to use LLVM's TTI or other downstream target descriptors that the IR knows nothing about (and doesn't need to).

I believe we need both extensions. Both are directly derivative of this PR and can be extended independently, so we really should not include it in this PR. Once we agree on the base structure, we'll continue with the more advanced strategies.

> My comment is about the case: `system1 = [deviceX[id = 0], deviceX[id = 1]]`, `system2 = [deviceX[id = 1], deviceX[id = 0]]`. In many cases `system1 == system2`, however, they are not the same by forcing ids. Moreover, it's also fragile, because a pass that works on `system1` might not work on `system2` because of the ordering.

This is only a problem if you don't know what device you're running and if you have more than one "system", which in the current PR design, is not reasonable. In the current design, _"system"_ is your **entire** system, including all devices in all nodes. Technically, having multiple systems with multiple devices is identical than having a single system with the same devices. Accessing them is slightly different, but equally unique (like array indexing `system[0][1]` vs `system0[1]`).

> Also, why do you need the ID to be stored in the device to query the properties? As far as I could tell, devices are placed in an array, so `attr.getEntries()[devId]` should work.

This means the order in which they appear in the IR is now relevant. It's a different design (means we need to sort maps before printing them out) but equally unique. In your multi-system example, `deviceX[id = 0]` is always the same device, regardless of which order it appears in the list and using some `getDeviceX(0)` would always return the one with `id = 0`, so there's no ambiguity.

The reason we use explicit (numeric) IDs in this PR is that we also want to have text IDs but haven't got around implementing it. 

This would help in many ways, for example:
* Generically naming the targets (ex. `CPU`, `GPU`) for simple systems.
* Using some device ID that is relevant to the driver (ex. `0xDEADBEEF`) to easily fetch properties via API calls.
* Encode information in the device ID (ex. `IPU[0][1]`) which could lead you to the second tile on the first device.

> My issue here is that this doesn't scale. Instead, the generic device attribute should only use a key lookup to find the value. If someone wants to create a more specialized attribute that contains L1, L2... they should be the ones adding those methods.

Correct. This is part of the two solutions I mention above. This is a planned extension, but we need to agree on the base structure now, which is the single role of this PR. The attributes encoded so far are only _examples_ of what can be done.

We don't plan to expand this to every possible combination of target attributes out there. We plan to have those for building the system's hierarchy, then using the generic entry type for target specific ones.

In summary, the answer to all your questions is simply: Yes, we agree with you, but this is a long walk and we've just started. We need to agree on the usage of DLTI and the base strategy for the system descriptor first. IF we do, then we continue implementing the rest in following PRs. 

However, some questions will only be answered when we start using the infra, and this is why we're doing step-wise. There's an issue from Intel's graph-compiler that links here, and that's one of the users. We hope others will have equivalent work, and we need to evolve together. 

It would be wrong to come up with a fixed _"generic"_ design right now, trying to predict all cases, and having no one actually using any of those. It's much simpler and faster to agree on a base design and improve from there, with the input of all users, not a single one.

fabianmcg commented:
Thanks for the replies @rengolin! Now I understand a little better your rationale. Btw, none of my comments are blocking I think it's a good implementation, so if you already have consensus go ahead.

The summary of my point is: the generic attributes and interfaces could be simpler and in consequence more generic, leaving all the real burden to derived attributes:

> It would be wrong to come up with a fixed _"generic"_ design right now, trying to predict all cases, and having no one actually using any of those. It's much simpler and faster to agree on a base design and improve from there, with the input of all users, not a single one.

I 100% agree, that's why I think we shouldn't include the `getSpecForMaxVectorOpWidth`, `getSpecForL1CacheSizeInBytes` in the interface nor in the generic attribute. Those belong in derived target attributes.

> I believe we need both extensions. Both are directly derivative of this PR and can be extended independently, so we really should not include it in this PR. Once we agree on the base structure, we'll continue with the more advanced strategies.

I think another solution for this is consolidating the `TargetDeviceSpecInterface` and `TargetSystemSpecInterface` in a single `TargetSpecInterface`.  And having something similar to:
```
struct Target {
  TargetList subTargets;
  PropertyMap properties;
};
```
Or even making properties include `subTargets`.

In this case, the interface only has `getSpecProp(StringRef)`, `getDeviceList()`, which would cover all target description cases without requiring extra changes in future PRs. Making things like: 
```mlir
// Using generic target attribute:
#Node = #dlti.target</*subTargets=*/[
  #rocdl.target<chip="gfx90a">,
  // Generic attribute:
  #dlti.target</*properties=*/{kind = "CPU", model = "4750U", vendor = "AMD"}>
  ], {topology = #numa.topology<...>}>
// Derived target attribute:
#cluster.target<
  node = #Node,
  num_nodes = 4096,
  interconnect = #fabric.infiniband<NDR>
>
```
Implementable after this PR.

> It's a different design (means we need to sort maps before printing them out) but equally unique.

This could create issues because nothing prevents creating two attributes with the same unordered device list but with a different list order (ie. [dev[id=0], dev[id=1]], [dev[id=1], dev[id=0]]). Those attributes wouldn't be equal at the time of creation, but equal after printing and parsing.

If we want to impose an order on the devices or provide identifiers, I still don't see how adding ID's to the targets is needed or improves the design. 
For this I think it's better doing a `DictAttr`, as the correspondence between ID and device is unambiguous for the developer and in the IR.

joker-eph commented on mlir/include/mlir/Dialect/DLTI/DLTIAttrs.td:
```diff
@@ -0,0 +1,207 @@
+    /// Since L1 cache size is an optional property, this function will
+    /// return a valid spec if the property is defined, otherwise it
+    /// will return an empty spec.
+    DataLayoutEntryInterface getSpecForL1CacheSizeInBytes();
```
Still have specialized APIs here? (L1 and vector width)

nhasabni replied:
Removed these APIs now.

joker-eph commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -194,6 +194,116 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
+                      "present. Otherwise, return empty spec.",
+      /*retTy=*/"::mlir::DataLayoutEntryInterface",
+      /*methodName=*/"getSpecForL1CacheSizeInBytes",
+      /*args=*/(ins)
```
Still L1 and vector width API here, is this intended? If so why?

joker-eph commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -362,6 +479,28 @@ def DataLayoutOpInterface : OpInterface<"DataLayoutOpInterface"> {
+      /*methodBody=*/"",
+      /*defaultImplementation=*/[{
+        return ::mlir::detail::getL1CacheSizeInBytes(entry);
+      }]
```
(same here)

joker-eph commented on mlir/lib/Dialect/DLTI/DLTI.cpp:
```diff
@@ -277,51 +236,179 @@ DataLayoutSpecAttr::combineWith(ArrayRef<DataLayoutSpecInterface> specs) const {
+    } else if (entryName == DLTIDialect::kTargetDeviceL1CacheSizeInBytesKey) {
+      IntegerAttr value =
+          llvm::dyn_cast_if_present<IntegerAttr>(entry.getValue());
+      if (!value || !value.getType().isUnsignedInteger(32))
```
Right: would want the IR to be `#dlti.dl_entry<"dlti.L1_cache_size_in_bytes", 4096>>` I think.

nhasabni replied:
I believe this is because the type of return value of a property query API needs to match with the type of the value right? E.g., the return value type of `getL1CacheSizeInBytes(DataLayoutEntryInterface)` is fixed to `std::optional<uint32_t>`, so the property value needs to be `uint32_t`, right? Or am I missing something?

joker-eph replied:
I would change `getL1CacheSizeInBytes` to simply use `int32_t` (reserve unsigned int for bit fields and other bit manipulation, prefer using signed int for normal quantities).

joker-eph commented:
> For this I think it's better doing a DictAttr, as the correspondence between ID and device is unambiguous for the developer and in the IR.

I think this is implemented now? The system is a dictionary of StringAttr key for the device ID to a device attribute (which does not contain the ID).

rengolin commented:
> > For this I think it's better doing a DictAttr, as the correspondence between ID and device is unambiguous for the developer and in the IR.
> 
> I think this is implemented now? The system is a dictionary of StringAttr key for the device ID to a device attribute (which does not contain the ID).

I think so. As I wrote this I realised, but was multi-threading so pushed the comment without update. :D

fabianmcg commented:
> I think this is implemented now? The system is a dictionary of StringAttr key for the device ID to a device attribute (which does not contain the ID).

You're right, I was looking at https://github.com/llvm/llvm-project/commit/85663ae6ccc57fe9b17fad7cad5d53a665755e7f , and I missed the new commits.

nhasabni commented:
> > For this I think it's better doing a DictAttr, as the correspondence between ID and device is unambiguous for the developer and in the IR.
> 
> I think this is implemented now? The system is a dictionary of StringAttr key for the device ID to a device attribute (which does not contain the ID).

Yes.

nhasabni committed:
Addressing review comments

nhasabni committed:
Addressing review comments

rengolin commented:
@ftynse are all your questions addressed?

ftynse commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -640,6 +677,46 @@ uint64_t mlir::DataLayout::getStackAlignment() const {
+  checkValid();
+  DataLayoutEntryInterface entry;
+  if (originalTargetSystemDesc) {
+    if (auto device =
```
Nit: expand auto unless the type is obvious.

ftynse commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -744,6 +821,56 @@ LogicalResult mlir::detail::verifyDataLayoutSpec(DataLayoutSpecInterface spec,
+    }
+  }
+
+  for (const auto &kvp : device_desc_keys) {
```
Nit: `(auto &&[first, second] : ...` works now.

joker-eph replied:
Do you need 2 `&`?

ftynse commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -744,6 +821,56 @@ LogicalResult mlir::detail::verifyDataLayoutSpec(DataLayoutSpecInterface spec,
+LogicalResult
+mlir::detail::verifyTargetSystemSpec(TargetSystemSpecInterface spec,
+                                     Location loc) {
+  DenseMap<StringAttr, DataLayoutEntryInterface> device_desc_keys;
+  DenseSet<TargetSystemSpecInterface::DeviceID> device_ids;
```
MLIR uses camelBack names. Please fix here and elsewhere.

ftynse reviewed (approved):
LGTM with a bunch of stylistic nits.   Please also think about the overall design for future steps :)

joker-eph commented on mlir/include/mlir/Interfaces/DataLayoutInterfaces.td:
```diff
@@ -194,6 +194,116 @@ def DataLayoutSpecInterface : AttrInterface<"DataLayoutSpecInterface"> {
+      /*args=*/(ins),
+      /*methodBody=*/""
+    >,
+  ];
```
The interface still has the L1 size and vector width methods, didn't we agree these would be removed everywhere?

nhasabni replied:
Do you mean we do not want any method for L1 size and vector width in this PR? What would be the interface/API to retrieve values of these keys from a device specification?

rengolin replied:
I'm not sure how the max vector size helps, with not all ops being supported in all vector sizes, etc.

For the L1, the problem is that we don't have any other target feature to show it works in a test.

How about, in the spirit of data layout, we have a string of cache levels instead? A shared format that describes L1/2/3 in CPUs as well as Thread/Warp/Core memory access for GPUs?

That'd be a generic format that we can encode in a string and work across various targets, and not be a specific property of CPUs.

joker-eph replied:
> Do you mean we do not want any method for L1 size and vector width in this PR? What would be the interface/API to retrieve values of these keys from a device specification?

Shouldn't everything be string based here?

rengolin replied:
At this level, I agree everything should be a string. Either a single value, an encoded pattern or the name of a target that will be retrieved from LLVM TTI or some internal handler (from a JSON file, for example). The implementation details for those should be in future PRs.

joker-eph commented on mlir/lib/Dialect/DLTI/DLTI.cpp:
```diff
@@ -332,11 +291,134 @@ DataLayoutSpecAttr DataLayoutSpecAttr::parse(AsmParser &parser) {
+
+StringAttr TargetDeviceSpecAttr::getL1CacheSizeInBytesIdentifier() {
+  return Builder(getContext())
+      .getStringAttr(DLTIDialect::kTargetDeviceL1CacheSizeInBytesKey);
```
This hits the context unnecessarily, the dialect itself should have a cached copy of the attribute.

Anyway: can we just finally prune this PR from any method that has to do with L1 and vector width?

joker-eph commented on mlir/lib/Dialect/DLTI/DLTI.cpp:
```diff
@@ -375,32 +457,34 @@ class TargetDataLayoutInterface : public DataLayoutDialectInterface {
+    for (DataLayoutEntryInterface dl_entry : entry.getEntries()) {
+      StringRef entryName = dl_entry.getKey().get<StringAttr>().strref();
+      // Check that the key name is known to us. Although, we may allow keys
+      // unknown to us.
```
What is the logic behind this verifier? The comment says "we may allow keys unknown to us", but we don't right now?

(also we should avoid string comparisons)

nhasabni replied:
Clarifying the comment - we don't allow unknown keys currently.

joker-eph commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -293,6 +293,28 @@ mlir::detail::getDefaultStackAlignment(DataLayoutEntryInterface entry) {
+// is empty (meaning the spec is missing), returns std::nullopt.
+std::optional<int64_t>
+mlir::detail::getMaxVectorOpWidth(DataLayoutEntryInterface entry) {
+  if (entry == DataLayoutEntryInterface())
```
```suggestion
  if (!entry)
```

But is this check useful?

joker-eph commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -324,6 +346,19 @@ static DataLayoutSpecInterface getSpec(Operation *operation) {
 
+static TargetSystemSpecInterface getTargetSystemSpec(Operation *operation) {
+  if (operation) {
+    ModuleOp moduleOp;
+    if (isa<ModuleOp>(operation)) {
+      moduleOp = llvm::dyn_cast<ModuleOp>(operation);
+    } else {
+      moduleOp = operation->getParentOfType<ModuleOp>();
+    }
```
```suggestion
    ModuleOp moduleOp = dyn_cast<ModuleOp>(operation);
    if (!moduleOp)
      moduleOp = operation->getParentOfType<ModuleOp>();
```

joker-eph commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -324,6 +346,19 @@ static DataLayoutSpecInterface getSpec(Operation *operation) {
+      moduleOp = operation->getParentOfType<ModuleOp>();
+    }
+    return moduleOp.getTargetSystemSpec();
+  } else
```
No else after return 

```suggestion
  }
```

nhasabni committed:
Addressing review comments - removing methods for L1 size and max vector width

nhasabni commented on mlir/lib/Interfaces/DataLayoutInterfaces.cpp:
```diff
@@ -744,6 +821,56 @@ LogicalResult mlir::detail::verifyDataLayoutSpec(DataLayoutSpecInterface spec,
+    // Ignore attributes that belong to an unknown dialect, the dialect may
+    // actually implement the relevant interface but we don't know about that.
+    if (!dialect)
+      continue;
```
I was following the logic from earlier parts of DLTI. I've fixed it though.

joker-eph reviewed (approved):
Thanks for the patience throughout the iterations @nhasabni !

nhasabni commented:
> Thanks for the patience throughout the iterations @nhasabni !

No problem! Thanks for the approval @joker-eph @ftynse!

rengolin commented:
FYI, here's our tracking of the following steps:
https://github.com/plaidml/tpp-mlir/issues/934
